### ISSUE TITLE
MemorySanitizer use-of-uninitialized-value error when running this code

### ISSUE THREAD

**Issue by gpollo** [rust-lang/rust#130038]
Running this code under MemorySanitizer returns an error. I do not know if it is a true positive or not. This code is copied from [`std::thread::Builder`](https://github.com/rust-lang/rust/blob/master/library/std/src/thread/mod.rs#L465-L467).

```rust
fn main() {
    let amt = std::env::var_os("RUST_MIN_STACK")
        .and_then(|s| s.to_str().and_then(|s| s.parse().ok()))
        .unwrap_or(100);
    println!("{}", amt);
}
```

```bash
$ RUSTFLAGS="-Z sanitizer=memory" cargo run --target x86_64-unknown-linux-gnu
...
SUMMARY: MemorySanitizer: use-of-uninitialized-value /rustc/94885bc699512cfee8560e73c2a01ee6b4b76563/library/core/src/option.rs:1444:9 in core::option::Option$LT$T$GT$::and_then::h8fffdb0e33c848c1
```

### Meta

``` bash
$ rustc --version --verbose
rustc 1.83.0-nightly (94885bc69 2024-09-01)
binary: rustc
commit-hash: 94885bc699512cfee8560e73c2a01ee6b4b76563
commit-date: 2024-09-01
host: x86_64-unknown-linux-gnu
release: 1.83.0-nightly
LLVM version: 19.1.0
```

**Comment by saethlin** [comment]
This is a known tripping hazard with MemorySanitizer: If any part of the program is not instrumented it tends to emit false positives. The precompiled standard library is not instrumented, so MemorySanitizer tends to assume all the memory in it is uninitialized when it gets exposed to the instrumented code that's compiled for the local crate.

The false positive is fixed by using `-Zbuild-std`, or by using `cargo-careful`: https://github.com/RalfJung/cargo-careful

**Comment by gpollo** [comment]
It does seems to fix the issue, thanks for clarifying this!
